### DISCUSSION TITLE
parse strings

### DISCUSSION THREAD

**Discussion by mfulton26** [colinhacks/zod#953] [Ideas]
In the spirit of [parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/), I'd like Zod to return a number for `z.number().parse("42")` instead of throwing an error. I'd like the same for booleans, arrays, objects, basically anything that has a well adopted string representation via JSON as well as widely accepted formats like ISO date strings.

I want to be able to use Zod for parsing, not validating, incoming requests from a client and I want my server to be conservative in what it sends back to the client but be liberal in what it accepts ([robustness principle](https://en.m.wikipedia.org/wiki/Robustness_principle)).

**Comment by villesau** [comment]
You can achieve it with `preprocess`:

```ts
z.preprocess(
  a => parseInt(z.string().parse(a), 10),
  z.number()
);
```

Though, I'd like to see something a bit less verbose and cumbersome to achieve the same. Something like `z.number().positive().parseFrom(z.string())` or `z.number().positive().coarseFrom(val => parseInt(z.string().parse(val), 10))` (but in simpler manner) or similar. I was also pretty surprised that the library does not easily support parsing as it is normally understood in JS ecosystem.

**Comment by killercup** [comment]
I too was quite surprised that parsing from strings is not a built-in feature. What made this whole experience way more annoying was that `zod.preprocess(parseNumber, zod.number())` is no longer a ZodNumber, and thus does not have methods like `.gt` and `.positive`. I understand why that is the case right now, but that really makes for a bad surprise.

(I even tried create a subclass of ZodNumber that overwrites its `_parse` method, but as `zod.number()` is actually calling the static `ZodNumber.create` which in turn calls its own constructor, the only way out is weirdly hacking the returned objects prototype… so I accept the bad API for now 😅 )

**Comment by KaKi87** [comment]
Hello, I also need this to parse numbers and booleans searchParams which are always parsed as strings. Thanks

**Reply by KaKi87** [reply]
> @JacobWeisenburger converted this issue into discussion

Why ? From my point of view, this is a feature request.
Thanks

**Reply by JacobWeisenburger** [reply]
I think you might be thinking that parse takes a value and returns another value. Common mistake. But in zod, parse doesn't do this. You need to define a [`preprocess`](https://github.com/colinhacks/zod#preprocess) or [`transform`](https://github.com/colinhacks/zod#transform) on your schema to do this.

**Reply by mfulton26** [reply]
@JacobWeisenburger if I understand correctly, you are saying that, by default, Zod validates instead of parses. This is very inconvenient. I was hoping to use this library. I think I'll continue to use [Joi](https://joi.dev/). Its schemas are more [robust](https://en.m.wikipedia.org/wiki/Robustness_principle) as they, by default, [parse instead of validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).

**Reply by mfulton26** [reply]
If Zod were to parse strings into types by default then it would be useful in a server context where I need to parse incoming data.

I created this issue as a feature request. As neat as preprocess/transform solutions are they require me to now write more code, etc.

I'd like to see Zod have built-in support for well understood string representations of types.

Thank you.

**Reply by KaKi87** [reply]
> I think I'll continue to use [Joi](https://joi.dev/)

Personally I don't have the luxury of choice for a [Deno](https://deno.land/) project.

**Reply by mfulton26** [reply]
Joi works in a browser so I don't see any reason it wouldn't work in Deno via skypack, jspm.io, etc.

Deno doesn't require all dependencies to be written for Deno only that they support standard APIs like in a browser and use ES Modules.

**Reply by KaKi87** [reply]
Damn, I already use JSPM for some stuff but I didn't realized that Joi supports browsers. You just freed me, thanks !

**Reply by JacobWeisenburger** [reply]
> @JacobWeisenburger if I understand correctly, you are saying that, by default, Zod validates instead of parses...

Zod is a ... **validation** library

https://github.com/colinhacks/zod/blob/c63a5988613f0accb2099d88f05db4201618ad6e/README.md?plain=1#L107-L109

---
> If Zod were to parse strings into types

zod's core purpose is not to parse strings, it is to validate schemas. I know it's confusing, because zod chose to use the word parse to mean validate, but that's just how zod did it.

---
If you would like me to convert this back into an issue, I will. But zod is very unlikely to change the meaning of `.parse()`

**Reply by KaKi87** [reply]
Alright, I understand, but yes, this vocabulary is confusing, perhaps you could explain it in the README.

**Reply by mfulton26** [reply]
Why does it reference "parse, don't validate"?

https://github.com/colinhacks/zod/blob/c63a5988613f0accb2099d88f05db4201618ad6e/README.md?plain=1#L120

**Reply by mfulton26** [reply]
If Zod is meant as only a validation library then I think that greatly limits its usefulness. If it is meant to be a go-to library for taking user/client supplied runtime values and coercing them into type-safe values to ease writing robust code then I think it needs to add rich support for common string representations. I thought/hoped Zod was meant to be TypeScript-first alternative to Joi but based on this discussion I don't think Zod that is the case. Zod uses schemas to validate. Joi uses schemas to parse. Parsing > Validating for my (most?) purposes.

I suggest updating the readme to not reference parsing if Zod isn't supposed to take such a functional approach. Thanks!

**Comment by JacobWeisenburger** [comment]
Use [`preprocess`](https://github.com/colinhacks/zod#preprocess)


Below also addresses the comment from @killercup
> ... and thus does not have methods like `.gt` and `.positive` ...
```ts
const positiveNumberFromStringOrNumber = z.preprocess(
    input => {
        if ( typeof input === 'number' ) return input
        if ( typeof input === 'string' ) return parseFloat( input )
    },
    z.number().gt( 0 ) // you can add string validators here
)

console.log( positiveNumberFromStringOrNumber.safeParse( 1 ).success ) // true
console.log( positiveNumberFromStringOrNumber.safeParse( '1' ).success ) // true

console.log( positiveNumberFromStringOrNumber.safeParse( 0 ).success ) // false
console.log( positiveNumberFromStringOrNumber.safeParse( '0' ).success ) // false

console.log( positiveNumberFromStringOrNumber.safeParse( -1 ).success ) // false
console.log( positiveNumberFromStringOrNumber.safeParse( '-1' ).success ) // false
```

**Reply by KaKi87** [reply]
Ultimately, both your proposal and `Joi.number()` accept `"123"` and parse it as `123`, which is expected.
But, your proposal also accepts `"123abc"` and parses it as `123`, which is a problem.

**Reply by JacobWeisenburger** [reply]
My answer (https://github.com/colinhacks/zod/discussions/953#discussioncomment-2223525) was addressing the original question in this thread. If you have a specific question that you would like answered, please start a [new discussion](https://github.com/colinhacks/zod/discussions/new) with your information and I will get to it as soon as I can. Thanks for your interest in zod.

**Reply by KaKi87** [reply]
My request is the same as the original :
> I'd like Zod to return a number for `z.number().parse("42")` instead of throwing an error. I'd like the same for booleans, arrays, objects

But, I need the parsing to be as strict as if the variable's actual type was the targeted type.
Thanks

*(Sorry for sending my reply twice from the wrong account :sweat_smile:)*

**Reply by JacobWeisenburger** [reply]
Are you saying that you are @mfulton26? I'm really confused.

---

please see my reply to mfulton26:
https://github.com/colinhacks/zod/discussions/953#discussioncomment-2229777

**Reply by KaKi87** [reply]
> Are you saying that you are @mfulton26?

No, I am not :sweat_smile:
But if you'd have email notifications enabled you'd see that I have sent this message three times in total, the two first times with another account, my professional one, which I am not supposed to use for that purpose.
I sometimes get confused between windows xD
